tree/ntuple/v7/inc/ROOT/RColumnElementBase.hxx

```diff
@@ -0,0 +1,166 @@
+ * For the list of contributors see $ROOTSYS/README/CREDITS.             *
+ *************************************************************************/
+
+#ifndef ROOT7_RColumnElementBase
+#define ROOT7_RColumnElementBase
+
+#include "RtypesCore.h"
+#include <ROOT/RError.hxx>
+#include <ROOT/RFloat16.hxx>
+#include <ROOT/RNTupleUtil.hxx>
+
+#include <Byteswap.h>
+#include <TError.h>
+
+#include <cstring> // for memcpy
+#include <cstddef> // for std::byte
+#include <cstdint>
+#include <memory>
+#include <string>
+#include <type_traits>
+#include <typeinfo>
+#include <utility>
+
+namespace ROOT::Experimental::Internal {
+
+// clang-format off
+/**
+\class ROOT::Experimental::Internal::RColumnElementBase
+\ingroup NTuple
+\brief A column element encapsulates the translation between basic C++ types and their column representation.
+
+Usually the on-disk element should map bitwise to the in-memory element. Sometimes that's not the case
+though, for instance on big endian platforms or for bools.
+
+There is a template specialization for every valid pair of C++ type and column representation.
+These specialized child classes are responsible for overriding `Pack()` / `Unpack()` for packing / unpacking elements
+as appropriate.
+*/
+// clang-format on
+class RColumnElementBase {
+protected:
+   /// Size of the C++ value that corresponds to the on-disk element
+   std::size_t fSize;
+   std::size_t fBitsOnStorage;
+
+   explicit RColumnElementBase(std::size_t size, std::size_t bitsOnStorage = 0)
+      : fSize(size), fBitsOnStorage(bitsOnStorage ? bitsOnStorage : 8 * size)
+   {
+   }
+
+public:
+   RColumnElementBase(const RColumnElementBase &other) = default;
+   RColumnElementBase(RColumnElementBase &&other) = default;
+   RColumnElementBase &operator=(const RColumnElementBase &other) = delete;
+   RColumnElementBase &operator=(RColumnElementBase &&other) = default;
+   virtual ~RColumnElementBase() = default;
+
+   /// If CppT == void, use the default C++ type for the given column type
+   template <typename CppT = void>
+   static std::unique_ptr<RColumnElementBase> Generate(EColumnType type);
+   static std::string GetTypeName(EColumnType type);
+   /// Most types have a fixed on-disk bit width. Some low-precision column types
+   /// have a range of possible bit widths. Return the minimum and maximum allowed
+   /// bit size per type.
+   static std::pair<std::uint16_t, std::uint16_t> GetValidBitRange(EColumnType type);
+
+   /// Derived, typed classes tell whether the on-storage layout is bitwise identical to the memory layout
+   virtual bool IsMappable() const
+   {
+      R__ASSERT(false);
+      return false;
+   }
+
+   /// If the on-storage layout and the in-memory layout differ, packing creates an on-disk page from an in-memory page
+   virtual void Pack(void *destination, const void *source, std::size_t count) const
+   {
+      std::memcpy(destination, source, count);
```

Comment:
These warning predate the PR. Still, can we do something about them? Can be a separate PR.

Reply:
I'll try to address them in a future PR, thanks for pointing them out.